Examples/Io/Csv/src/CsvBFieldWriter.cpp

```diff
@@ -0,0 +1,261 @@
+   */
+  std::array<std::size_t, ConfigType::NDims> bins;
+  Vector min, max;
+
+  if constexpr (Grid) {
+    /*
+     * First, we should check whether the Bfield actually has a
+     * three-dimensional bin count and size.
+     */
+    if (bins.size() != 3 || field.getMin().size() != 3 ||
+        field.getMax().size() != 3) {
+      throw std::invalid_argument(
+          "Input B-field does not support 3-dimensional indexing.");
+    }
+
+    /*
+     * If our magnetic field is grid-like, we know that it has a built-in size
+     * and bin count. We can use these, but the user may want to override the
+     * values with custom values.
+     */
+    for (std::size_t i = 0; i < ConfigType::NDims; ++i) {
+      /*
+       * For each dimension, we first get the corresponding value from the
+       * grid-based vector field...
+       */
+      bins[i] = field.getNBins()[i];
+      min[i] = field.getMin()[i];
+      max[i] = field.getMax()[i];
+
+      /*
+       * ...and then we override them with the optional user-supplied values.
+       */
+      if (config.bins[i]) {
+        bins[i] = *config.bins[i];
+      }
+
+      if (config.range[i][0]) {
+        min[i] = *config.range[i][0];
+      }
+
+      if (config.range[i][1]) {
+        max[i] = *config.range[i][1];
+      }
+    }
+  } else {
+    /*
+     * If the vector field is not grid based, then there is no side and we are
+     * forced to use the user-supplied data. Remember that in this case, the
+     * values are not optional.
+     */
+    for (std::size_t i = 0; i < ConfigType::NDims; ++i) {
+      bins[i] = config.bins[i];
+      min[i] = config.range[i][0];
+      max[i] = config.range[i][1];
+    }
+  }
+
+  /*
+   * Next, we calculate the size (in physical space) of each bin.
+   */
+  Vector delta;
+
+  for (std::size_t i = 0; i < ConfigType::NDims; ++i) {
+    delta[i] = (max[i] - min[i]) / bins[i];
+  }
+
+  /*
+   * Create the appropriate magnetic field context and cache to interact with
+   * the B fields.
+   */
+  Acts::MagneticFieldContext mctx{};
+  typename FieldType::Cache cache = field.makeCache(mctx);
+
+  /*
+   * Finally, we can begin to fill the output file with data from our B field.
+   * Again, the procedure is slightly different depending on whether we are
+   * working with Cartesian or cylindrical coordinates.
+   */
+  if constexpr (Coord == CoordinateType::XYZ) {
+    ACTS_INFO("Writing XYZ field of size "
+              << (bins[0] + 1) << " x " << (bins[1] + 1) << " x "
+              << (bins[2] + 1) << " to file " << config.fileName << "...");
+
+    std::size_t total_items = (bins[0] + 1) * (bins[1] + 1) * (bins[2] + 1);
+
+    /*
+     * For Cartesian coordinates, iterate over bins in the x, y, and z
+     * directions. Note that we iterate one additional time because we are
+     * writing the _edges_ of the bins, and the final bin needs to be closed.
+     */
+    for (std::size_t x = 0; x <= bins[0]; ++x) {
```

Comment:
now that I see this I wonder if there was a similar reason for `+1` and `<=` in https://github.com/acts-project/acts/pull/1381/files#diff-8195d95c9e69fcf3f1f7fade7f05b8253792f151a35fa7fc47da3c8800fc90b6L159 @timadye ?

Reply:
Good question, to me it makes sense that a 300 × 300 × 300 _bin_ B-field would consist of 301 × 301 × 301 vectors, because you need to account for the additional edge points. But we will wait to see what Tim thinks.

Reply:
Yeah, I removed the final bin to match the existing ROOT format. That gives field points at the lower-left point of each bin, not including the overflow bin. For the ATLAS B-field that I tested it on, it has 201 x 201 x 301 points, but that translates to 201 x 201 x 301 bins (not counting underflow and overflow bins).

For your CSV format, you can write whatever format you like, but it might be useful to keep consistent if possible. If you do include the overflow bin, then you should probably also include the underflow bin. Also, there may be problems calling `getField()` for value >= max (this is something I am trying to address in #1467).

Reply:
Hmm, I configured the output to be the same as the CSV files at https://gitlab.cern.ch/acts/acts-data/-/tree/master/MagneticField/ATLAS. Perhaps we should try to converge on a preferred format for this.

Reply:
Can you check how many bins are given in the `bins` array for the ATLAS B-field? When I last tested this, reading in `acts-itk/bfield/ATLAS-BField-xyz.root`, I got `bins = field.getNBins() = {201,201,301}`, so `x < bins[0]` should include all 201 elements in the CSV.

Sorry if I misunderstood.

Reply:
Hi Tim, I think perhaps the core problem here is that we disagree on what the concept of bin is. I think this makes sense in part because the term "bin" is somewhat undescriptive. Let's take, for example, the following example:

<img src="https://user-images.githubusercontent.com/1165820/193574109-5716fcf7-123e-4dc5-8cdd-e2a7dfa49297.png" width="200"/>

We might imagine this being some two-dimensional vector field spanning the range [-30mm, 30mm]<sup>2</sup>. It is my understanding that in such a model, the vector field has 49 "bins"; 7 in the horizontal axis and 7 in the vertical axis. Each of the bins is positioned exactly on the cross points of the horizontal and vertical lines, much like if one was playing a game of baduk. The distance between the points, in this case, is 10mm; this meshes nicely with the idea that we need 7 points to span a range of 60mm (which is 6 times the "bin" resolution if 10mm): it's the number of squares ("texels") plus one.

The only other intuition I can muster is to imagine that the "bins" are positioned at the center of the white squares defined by this grid. Perhaps this is the way you are thinking about this, and perhaps this is why we disagree on the structure of the output file?

The [ATLAS magnetic field in the Acts data repository](https://gitlab.cern.ch/acts/acts-data/-/tree/master/MagneticField/ATLAS) has a resolution of 10cm and spans 20 meters in the _x_ and _y_ axes, and 30 meters in the _z_ axis. The only way I can make sense of this is by imagining the "bins" as I described earlier: as the intersections of the grid lines. This also meshes with the contents of the file, which has increments of 10cm starting from the origin 0, rather than having the same grid but offset by 5cm, as would be expected in the alternative interpretation. The ATLAS field also has 201×201×301 "bins", rather than 200×200×300; the former is what would be produced by the implementation in this PR.

Reply:
Hi Stephen,

Thanks for the nice discussion. I always find this confusing, so each time have to go back to the source to understand it again. I'm pretty sure there is still an inconsistency, but I thought I should check my understanding by running the code.

For me, the best test would be to read in [`acts-itk/bfield/ATLAS-BField-xyz.root`](https://gitlab.cern.ch/atlas/acts-itk/-/tree/main/bfield) (or [`acts-data/MagneticField/ATLAS/ATLASBField_xyz.root`](https://gitlab.cern.ch/acts/acts-data/-/blob/master/MagneticField/ATLAS) if you prefer) and write it out as CSV. This should be possible with Python `acts.examples.MagneticFieldMapXyz` and your new `acts.examples.CsvBFieldWriter`. The CSV output should specify the same points as the ROOT input, or the `ATLASBField_xyz.txt` file that was used to create it.

Unfortunately, I can't get this to work. I modified a RootBFieldWriter example (adapted from [`bfield_writing.py`](https://github.com/acts-project/acts/blob/main/Examples/Scripts/Python/bfield_writing.py)) like this:

```
import acts.examples
field = acts.examples.MagneticFieldMapXyz("acts-itk/bfield/ATLAS-BField-xyz.root")
cfg = acts.examples.CsvBFieldWriter.ConfigXyzGrid()
cfg.bField = field
cfg.fileName = str("ATLAS-BField-xyz.csv")
acts.examples.CsvBFieldWriter.runXyzGrid(cfg, acts.logging.VERBOSE)
```

It's a bit unfortunate that the Csv interface is different from the Root interface. Can that be simplified?

Anyway, my attempt gives:

```
RuntimeError: B-field returned a non-extant value!
```

Do you know what's going on?

Thanks,
Tim.

Reply:
I just noticed that it does print a message:

```
15:49:51    CsvBFieldWri   INFO      Writing XYZ field of size 202 x 202 x 302 to file ATLAS-BField-xyz.csv...
```
and created the CSV file, but only the first row with `(x,y,z) = (-10000,-10000,-15000...14900.33)`. That does indeed have the wrong z bins, but of course this isn't a complete test.